src/common/librouter/test/usock_epipe.c

```diff
@@ -0,0 +1,255 @@
+{
+    const struct flux_msg_cred *cred;
+
+    cred = usock_conn_get_cred (conn);
+
+    usock_conn_set_error_cb (conn, server_error_cb, arg);
+    usock_conn_set_recv_cb (conn, server_recv_cb, arg);
+    usock_conn_set_close_cb (conn, server_close_cb, arg);
+
+    usock_conn_accept (conn, cred);
+}
+
+static int server_cb (flux_t *h, void *arg)
+{
+    flux_reactor_t *r = flux_get_reactor (h);
+    char sockpath[PATH_MAX + 1];
+    struct usock_server *server;
+
+    if (snprintf (sockpath,
+                  sizeof (sockpath),
+                  "%s/server",
+                  tmpdir) >= sizeof (sockpath)) {
+        diag ("usock_server_create buffer overflow");
+        return -1;
+    }
+    if (!(server = usock_server_create (r, sockpath, 0644))) {
+        diag ("usock_server_create failed");
+        return -1;
+    }
+    usock_server_set_acceptor (server, server_acceptor, arg);
+
+    if (flux_reactor_run (r, 0) < 0) {
+        diag ("flux_reactor_run failed");
+        return -1;
+    }
+    usock_server_destroy (server);
+    return 0;
+}
+
+/* End Test Server
+ */
+
+/* Send a burst of count small messages and closes connection.
+ * Assumes that the OS socket buffer is sufficient to contain all of it.
+ */
+static void test_send_and_exit (flux_t *h, int count)
+{
+    int i;
+    char sockpath[PATH_MAX + 1];
+    flux_msg_t *msg;
+    flux_msg_t *nmsg;
+    int fd;
+    struct usock_client *client;
+
+    if (!(msg = flux_request_encode ("init", NULL))
+        || !(nmsg = flux_request_encode ("nil", NULL)))
+        BAIL_OUT ("flux_request_encode failed");
+    if (flux_msg_pack (msg, "{s:i}", "expected", count) < 0)
+        BAIL_OUT ("flux_msg_pack failed");
+
+    if (snprintf (sockpath,
+                  sizeof (sockpath),
+                  "%s/server",
+                  tmpdir) >= sizeof (sockpath))
+        BAIL_OUT ("buffer overflow");
+    fd = usock_client_connect (sockpath, USOCK_RETRY_DEFAULT);
+    ok (fd >= 0,
+        "usock_client_connect %s works", sockpath);
+    ok ((client = usock_client_create (fd)) != NULL,
+        "usock_client_create works");
+
+    ok (usock_client_send (client, msg, 0) == 0,
+        "usock_client_send init message works: expected=%d", count);
+
+    for (i = 1; i < count; i++) {
+        ok (usock_client_send (client, nmsg, 0) == 0,
+            "usock_client_send[%d] works", i);
+    }
+
+    diag ("disconnecting");
+
+    usock_client_destroy (client);
+    (void)close (fd);
+    flux_msg_destroy (msg);
+}
+
+int main (int argc, char *argv[])
+{
+    struct test_params tp = {0};
+    flux_t *h;
+
+    plan (NO_PLAN);
+
+    tmpdir_create ();
+
+    signal (SIGPIPE, SIG_IGN);
+
+    diag ("starting test server");
+    test_server_environment_init ("usock_server");
+
+    if (!(h = test_server_create (server_cb, &tp)))
+        BAIL_OUT ("test_server_create failed");
+
+    pthread_mutex_lock (&server_mutex);
+    test_send_and_exit (h, 1);
+    pthread_mutex_lock (&server_mutex);
```

Comment:
Is it the case that the client thread only locks the mutex and the server thread only unlocks the mutex?  So the client thread is expected to block trying to retake the mutex?  I think that's a problem...

From [pthread_mutex_settype(3)](https://linux.die.net/man/3/pthread_mutexattr_settype):

> PTHREAD_MUTEX_DEFAULT
> Attempting to recursively lock a mutex of this type results in undefined behavior. Attempting to unlock a mutex of this type which was not locked by the calling thread results in undefined behavior. Attempting to unlock a mutex of this type which is not locked results in undefined behavior. An implementation may map this mutex to one of the other mutex types.

Might want a condition variable here.

Reply:
Bah, you're right. I was trying to be lazy as possible and failed.

Reply:
We should all give being lazy our best effort :-)

Reply:
Ok I pushed a fixup switching to a condition variable. Hopefully I did it correctly as it has been awhile.

It occurred to me that using pthread synchronization is not the zeromq nor "Flux" way to do things between client and server. Ideally the server would store test results that could then be "fetched" by the client (test main) to check whether any messages were dropped.